### DISCUSSION TITLE
Whitelisting email recipients (broke with Quarkus 3.0.0.Final)

### DISCUSSION THREAD

**Discussion by canidae** [quarkusio/quarkus#32685] [Community]
Prior to Quarkus 3.0.0.Final I could extend MutinyMailerImpl to implement a whitelisting of email recipients. Unfortunately, a [recent change](https://github.com/quarkusio/quarkus/commit/fb2affe542af486181755282763023c411865a42) broke this, and I don't see an easy way to reimplement my whitelisting with the latest iteration of the mailer.

This was implemented as a library used by 20 or so internal projects who all simply would `@Inject ReactiveMailer` as if they used `quarkus-mailer` directly, but with a whitelisting feature. Whitelisting is a useful feature for us, as we'd like to limit who we may send mails to during QA-testing (where the mock mailbox is not sufficient as we'd still like to send mails to some select recipients, but prevent accidentally sending to i.e. customers due to inadequately washed test data).

Preferably I'd like to see whitelisting as a feature in `quarkus-mailer`, but it seems like this isn't a feature requested by many other people. Thus my question is if I can easily modify the code below to work with Quarkus 3.0.0.Final.

Code:
```java
@Alternative
@Priority(1)
@ApplicationScoped
public class WhitelistMailerImpl extends MutinyMailerImpl {
    @Inject
    Logger log;

    @ConfigProperty(name = "mailer.whitelist.enabled")
    Optional<Boolean> whitelistEnabled;

    @ConfigProperty(name = "mailer.whitelist.emails")
    Optional<List<String>> whitelistEmails;

    @Override
    public Uni<Void> send(Mail... mails) {
        if (!whitelistEnabled.orElse(!LaunchMode.current().isDevOrTest()))
            return super.send(mails);

        var allowedMails = new ArrayList<Mail>();
        for (var mail : mails) {
            var to = filterWhitelisted(mail.getTo());
            var cc = filterWhitelisted(mail.getCc());
            var bcc = filterWhitelisted(mail.getBcc());
            if (to.isEmpty() && cc.isEmpty() && bcc.isEmpty()) {
                log.warnf("A mail to %s, cc %s, bcc %s with subject '%s' was not sent. Whitelisting is enabled and there were " +
                        "no whitelisted email addresses in the mail.", mail.getTo(), mail.getCc(), mail.getBcc(), mail.getSubject());
            } else {
                mail.setTo(to);
                mail.setCc(cc);
                mail.setBcc(bcc);
                allowedMails.add(mail);
            }
        }
        return allowedMails.isEmpty() ? Uni.createFrom().voidItem() : super.send(allowedMails.toArray(Mail[]::new));
    }

    private List<String> filterWhitelisted(List<String> emails) {
        var allowedRecipients = whitelistEmails.orElse(Collections.emptyList());
        var keep = new ArrayList<String>();
        for (var email : emails) {
            for (var regex : allowedRecipients) {
                if (email.matches(regex)) {
                    keep.add(email);
                    break;
                }
            }
        }
        return keep;
    }
}
```

**Comment by gsmet** [comment]
I created https://github.com/quarkusio/quarkus/issues/32688 to track the issue. Let's discuss things in the issue instead.
